pygmo/_py_algorithms.py

```diff
@@ -0,0 +1,131 @@
+import random
+
+try:
+    from scipy.optimize import minimize
+
+    class scipy:
+        """
+        This class is a user defined algorithm (UDA) providing a wrapper around the function scipy.optimize.minimize.
+        The constructor accepts those arguments that are specific to the algorithm:
+        - args
+        - method
+        - tol - the tolerance
+        - callback
+        - options
+
+        Other aspects, like bounds or the existence of a gradient and hessian, are taken from the problem.
+        """
+
+        def __init__(self, args=(), method=None, tol=None, callback=None, options=None):
+            method_list = [
+                "Nelder-Mead",
+                "Powell",
+                "CG",
+                "BFGS",
+                "Newton-CG",
+                "L-BFGS-B",
+                "TNC",
+                "COBYLA",
+                "SLSQP",
+                "trust-constr",
+                "dogleg",
+                "trust-ncg",
+                "trust-exact",
+                "trust-krylov",
+            ]
+            if method in method_list + [None]:
+                self.method = method
+            else:
+                raise ValueError("Method not supported: " + method)
+
+            self.args = args
+            self.tol = tol
+            self.callback = callback
+            self.options = options
+
+        def evolve(self, population):
+            """
+            Take a random member of the population, use it as initial guess
+            for calling scipy.optimize.minimize and replace it with the final result.
+
+            Modifies the given population and returns it.
+            """
+            problem = population.problem
+
+            if problem.get_nc() > 0:
```

Comment:
This is troublesome. Constraint are important for these type of solvers (there is not much they add to pygmo without). Could we discuss a way to actually include them?

Reply:
I'm currently working on a wrapper for them. Scipy requires a callable function for each constraint, while in pygmo all constraints are included in the fitness function. It would be easy to add a wrapper that calls the fitness function and then selects the right part of the result, but that would result in one additional function evaluation for each constraint. Thoughts?

Reply:
Since one output of the fitness function is going to be used by several other calls, I think it is best to store it in a wrapper class.

Reply:
uff really? no support for vector od constraints? what an unefficient logic. In similar cases in the past we implemented a caching mechanism ... so that if fitness(x) is called with the same x two times in a row it will just return the already computed values. But caches have their issues too ... and are prone to misuse ....

Reply:
I think there are fewer concerns in Python about caches, because in any case we are doing some form of multiprocessing for parallelism, and thus there is little danger of concurrent accesses to the same cache.

Reply:
Update on this: I've tried using the python caching mechanisms, but they require their inputs to be hashable, which numpy arrays are not. One could convert arrays to tuples as a workaround, but this also has a performance penalty.

Reply:
Caching of fitness values including the constraints is now implemented in 6a3d0f2, avoiding redundant function evaluations. I've used the simplest possible check, currently only ever one value is cached. This cache has no locks or precautions regarding thread safety, since the _fitness_wrapper object is constructed within the evolution method, so no two threads should have the same fitness_wrapper anyway.